packages/flutter/lib/src/widgets/snapshot_widget.dart

```diff
@@ -0,0 +1,434 @@
+///   performance of using this widget may regress performance due to the fact
+///   that both the UI and engine share a single thread.
+class SnapshotWidget extends SingleChildRenderObjectWidget {
+  /// Create a new [SnapshotWidget].
+  ///
+  /// The [controller] and [child] arguments are required.
+  const SnapshotWidget({
+    super.key,
+    this.mode = SnapshotMode.normal,
+    this.painter = const _DefaultSnapshotPainter(),
+    required this.controller,
+    required super.child
+  });
+
+  /// The controller that determines when to display the children as a snapshot.
+  final SnapshotController controller;
+
+  /// Configuration that controls how the snapshot widget decides to paint its children.
+  ///
+  /// Defaults to [SnapshotMode.normal], which throws an error when a platform view
+  /// or texture view is encountered.
+  ///
+  /// See [SnapshotMode] for more information.
+  final SnapshotMode mode;
+
+  /// The painter used to paint the child snapshot or child widgets.
+  final SnapshotPainter painter;
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    debugCheckHasMediaQuery(context);
+    return _RenderSnapshotWidget(
+      controller: controller,
+      mode: mode,
+      devicePixelRatio: MediaQuery.of(context).devicePixelRatio,
+      painter: painter,
+    );
+  }
+
+  @override
+  // ignore: library_private_types_in_public_api
+  void updateRenderObject(BuildContext context, covariant _RenderSnapshotWidget renderObject) {
+    debugCheckHasMediaQuery(context);
+    renderObject
+      ..controller = controller
+      ..mode = mode
+      ..devicePixelRatio = MediaQuery.of(context).devicePixelRatio
+      ..painter = painter;
+  }
+}
+
+// A render object that conditionally converts its child into a [ui.Image]
+// and then paints it in place of the child.
+class _RenderSnapshotWidget extends RenderProxyBox {
+  // Create a new [_RenderSnapshotWidget].
+  _RenderSnapshotWidget({
+    required double devicePixelRatio,
+    required SnapshotController controller,
+    required SnapshotMode mode,
+    required SnapshotPainter painter,
+  }) : _devicePixelRatio = devicePixelRatio,
+       _controller = controller,
+       _mode = mode,
+       _painter = painter;
+
+  /// The device pixel ratio used to create the child image.
+  double get devicePixelRatio => _devicePixelRatio;
+  double _devicePixelRatio;
+  set devicePixelRatio(double value) {
+    if (value == devicePixelRatio) {
+      return;
+    }
+    _devicePixelRatio = value;
+    markNeedsPaint();
+  }
+
+ /// The painter used to paint the child snapshot or child widgets.
+  SnapshotPainter get painter => _painter;
+  SnapshotPainter _painter;
+  set painter(SnapshotPainter value) {
+    if (value == painter) {
+      return;
+    }
+    final SnapshotPainter oldPainter = painter;
+    oldPainter.removeListener(markNeedsPaint);
+    _painter = value;
+    if (painter.shouldRepaint(oldPainter)) {
```

Comment:
if painter and oldPainter have different runtimetypes you probably always want to repaint without asking `shouldRepaint` similarly to how CustomPaint does it?

Reply:
Done